File: backend/src/routes/notifications.js
const express = require('express');
const prisma = require('../config/db');
const { auth } = require('../middleware/auth');

const router = express.Router();

// Get user notifications
router.get('/', auth, async (req, res, next) => {
  try {
    const { page = 1, limit = 20, unreadOnly } = req.query;
    const skip = (parseInt(page) - 1) * parseInt(limit);

    const where = { userId: req.user.id };
    if (unreadOnly === 'true') where.read = false;

    const [notifications, total, unreadCount] = await Promise.all([
      prisma.notification.findMany({
        where,
        orderBy: { createdAt: 'desc' },
        skip,
        take: parseInt(limit)
      }),
      prisma.notification.count({ where }),
      prisma.notification.count({ where: { userId: req.user.id, read: false } })
    ]);

    res.json({ notifications, total, unreadCount });
  } catch (error) {
    next(error);
  }
});

// Mark notification as read
router.put('/:id/read', auth, async (req, res, next) => {
  try {
    const notification = await prisma.notification.update({
      where: { id: req.params.id },
      data: { read: true }
    });
    res.json(notification);
  } catch (error) {
    next(error);
  }
});

// Mark all as read
router.put('/read-all', auth, async (req, res, next) => {
  try {
    await prisma.notification.updateMany({
      where: { userId: req.user.id, read: false },
      data: { read: true }
    });
    res.json({ message: 'All notifications marked as read' });
  } catch (error) {
    next(error);
  }
});

module.exports = router;

File: backend/app/routes/notifications.py
from datetime import datetime

from fastapi import APIRouter, Depends, Query
from sqlalchemy import func, select, update
from sqlalchemy.ext.asyncio import AsyncSession

from app.config.database import get_db
from app.middleware.auth import get_current_user
from app.models.models import Notification, User

router = APIRouter(prefix="/api/notifications", tags=["notifications"])


@router.get("/")
async def get_notifications(
    page: int = Query(1, ge=1),
    limit: int = Query(20, ge=1, le=100),
    unreadOnly: str | None = None,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    filters = [Notification.userId == user.id]
    if unreadOnly == "true":
        filters.append(Notification.read == False)

    query = (
        select(Notification)
        .where(*filters)
        .order_by(Notification.createdAt.desc())
        .offset((page - 1) * limit)
        .limit(limit)
    )
    result = await db.execute(query)
    notifications = result.scalars().all()

    total_r = await db.execute(select(func.count(Notification.id)).where(*filters))
    total = total_r.scalar()

    unread_r = await db.execute(
        select(func.count(Notification.id)).where(Notification.userId == user.id, Notification.read == False)
    )
    unread_count = unread_r.scalar()

    return {
        "notifications": [
            {
                "id": n.id,
                "type": n.type.value if hasattr(n.type, "value") else n.type,
                "title": n.title,
                "message": n.message,
                "read": n.read,
                "data": n.data,
                "createdAt": n.createdAt.isoformat(),
            }
            for n in notifications
        ],
        "total": total,
        "unreadCount": unread_count,
    }


@router.put("/{notification_id}/read")
async def mark_read(
    notification_id: str,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(select(Notification).where(Notification.id == notification_id))
    notification = result.scalar_one_or_none()
    if notification:
        notification.read = True
        await db.commit()
        await db.refresh(notification)
    return {
        "id": notification.id,
        "type": notification.type.value if hasattr(notification.type, "value") else notification.type,
        "title": notification.title,
        "message": notification.message,
        "read": notification.read,
        "data": notification.data,
        "createdAt": notification.createdAt.isoformat(),
    }


@router.put("/read-all")
async def mark_all_read(
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    await db.execute(
        update(Notification)
        .where(Notification.userId == user.id, Notification.read == False)
        .values(read=True)
    )
    await db.commit()
    return {"message": "All notifications marked as read"}
